internal/parser/translate.go

```diff
@@ -327,84 +327,155 @@ func fromIngressV1(log logrus.FieldLogger, ingressList []*networkingv1.Ingress)
+		hostnames := make([]*string, 0, len(spec.Hostnames))
+		for _, hostname := range spec.Hostnames {
+			hostnames = append(hostnames, kong.String(string(hostname)))
+		}
+
+		// each rule may represent a different set of backend services that will be accepting
+		// traffic, so we make separate routes and Kong services for every present rule.
 		for _, rule := range spec.Rules {
-			for _, match := range rule.Matches {
-				paths = append(paths, match.Path.Value)
+			// the HTTPRoute specification upstream specifically defines matches as
+			// independent (e.g. each match is an OR with other matches, not an AND).
+			// Therefore we treat each match rule as a separate Kong Route, so we iterate through
+			// all matches to determine all the routes that will be needed for the services.
+			routes := make([]kongstate.Route, 0, len(rule.Matches))
+			for matchNumber, match := range rule.Matches {
+				// determine the name of the route, identify it as a route that belongs
+				// to a Kubernetes HTTPRoute object.
+				routeName := kong.String(fmt.Sprintf(
+					"httproute.%s.%s.%d",
+					httproute.Namespace,
+					httproute.Name,
+					matchNumber,
+				))
+
+				// TODO: implement query param matches
+				if len(match.QueryParams) > 0 {
+					log.Errorf("regular expression path matches are not yet supported, HTTPRoute %s/%s can't be routed for match %+v", httproute.Namespace, httproute.Name, match)
+					continue
+				}
+
+				// TODO: implement regex path matches
+				if *match.Path.Type == gatewayv1alpha2.PathMatchRegularExpression {
+					log.Errorf("regular expression path matches are not yet supported, HTTPRoute %s/%s can't be routed for match %+v", httproute.Namespace, httproute.Name, match)
+					continue
+				}
+
+				// build the route object using the method and pathing information
+				r := kongstate.Route{
+					Ingress: objectInfo,
+					Route: kong.Route{
+						Name:         routeName,
+						Protocols:    kong.StringSlice("http", "https"),
+						PreserveHost: kong.Bool(true),
+						Hosts:        hostnames,
+					},
+				}
+				log.Debugf("generated route %s for HTTPRoute %s/%s", routeName, httproute.Namespace, httproute.Name)
+
+				// configuration path matching information about the route if paths
+				// matching was defined.
+				if match.Path != nil {
+					// determine the path match values
+					r.Route.Paths = []*string{match.Path.Value}
+
+					// determine whether path stripping needs to be enabled
+					r.Route.StripPath = kong.Bool(match.Path.Type == nil || *match.Path.Type == gatewayv1alpha2.PathMatchPathPrefix)
+				}
+
+				// configure method matching information about the route if method
+				// matching was defined.
+				if match.Method != nil {
+					r.Route.Methods = append(r.Route.Methods, kong.String(string(*match.Method)))
+				}
+
+				// convert header matching to from HTTPRoute to Route format
+				invalidHeaders := false
+				headers := make(map[string][]string)
```

Comment:
this conversion appears to be a pure mathematical function, so it probably make sense to extract it to a function and unit test it. Also, `err != nil` would be more idiomatic than `invalidHeaders`

Reply:
f19290b3e5674e9fd8247165d101d70c952ff537 resolves both of these